src/ItemVirtualMachine.php

```diff
@@ -203,45 +201,51 @@ public static function showForVirtualMachine(CommonDBTM $asset)
-                echo "<table class='tab_cadre_fixehov'>";
-                echo  "<tr class='noHover'><th colspan='2' >" . __('List of hosts') . "</th></tr>";
-
-                $header = "<tr><th>" . __('Name') . "</th>";
-                $header .= "<th>" . Entity::getTypeName(1) . "</th>";
-                $header .= "</tr>";
-                echo $header;
-
+                TemplateRenderer::getInstance()->display('components/form/item_virtualmachine.html.twig', [
+                    'has_host' => true,
+                ]);
                 $computer = new Computer();
+                $entries = $columns = [];
                 foreach ($hosts as $host) {
-                    $class = $host['is_deleted'] ? "deleted" : "";
-                    echo "<tr class='tab_bg_2 $class' >";
-                    echo "<td>";
-                    if ($computer->can($host['computers_id'], READ)) {
-                        echo "<a href='" . Computer::getFormURLWithID($computer->fields['id']) . "'>";
-                        echo $computer->fields['name'] . "</a>";
-                        $tooltip = "<table><tr><td>" . __('Name') . "</td><td>" . $computer->fields['name'] .
-                             '</td></tr>';
-                        $tooltip .= "<tr><td>" . __('Serial number') . "</td><td>" . $computer->fields['serial'] .
-                             '</td></tr>';
-                        $tooltip .= "<tr><td>" . __('Comments') . "</td><td>" . $computer->fields['comment'] .
-                             '</td></tr></table>';
-                        echo "&nbsp; " . Html::showToolTip($tooltip, ['display' => false]);
+                    if ($computer->can($host['items_id'], READ)) {
+                        $columns = [
+                            'name' => __('Name'),
+                            'serial' => __('Serial number'),
+                            'comment' => __('Comments'),
+                            'entity' => _n('Entity', 'Entities', 1)
+                        ];
+                        $entries[] = [
+                            'name' => $computer->getLink($host['items_id']),
+                            'serial' => $computer->fields['serial'],
+                            'comment' => $computer->fields['comment'],
+                            'entity' => Dropdown::getDropdownName('glpi_entities', $computer->fields['entities_id'])
+                        ];
                     } else {
-                        echo $computer->fields['name'];
+                        $columns = [
+                            'name' => __('Name'),
+                            'entity' => _n('Entity', 'Entities', 1)
+                        ];
+                        $entries[] = [
+                            'name' => $computer->fields['name'],
+                            'entity' => Dropdown::getDropdownName('glpi_entities', $computer->fields['entities_id'])
+                        ];
                     }
```

Comment:
I guess `$computer->can($host['items_id'], READ)` may return a different result for each row. In this case, you will have a mix of entries with 4 columns and entries with 2 columns. Also, `$columns` is redefined on each loop iteration.

IMHO, `$columns` should be defined once, and then every entries should contain the 4 columns (with empty values if they should ot be readable).